### DISCUSSION TITLE
Export Components using ESM

### DISCUSSION THREAD

**Discussion by yordis** [tailwindlabs/headlessui#852] [Ideas]
Hey there, I notice that you do the following as of today:

https://github.com/tailwindlabs/headlessui/blob/452b2c2852950cd846fe58e49ad432bb160e6e66/packages/%40headlessui-react/src/components/dialog/dialog.tsx#L421

Such things force me to use `Dialog` object, I have seen such practices before and they always backfire on us due to extra complexity re-exporting things or copying over, and tree-shaking (no problem as of today, yet).

Given the context that ESM is already an object if people really want a single import `import * as Something ...`

----

Recently I had to wrap `Dialog` since I had to add FullStory `className` by default.

```tsx
import * as React from 'react';
import { Dialog as HeadlessDialog } from '@headlessui/react';
import clsx from 'classnames';
import { FULLSTORY_UNMASK } from '@/styles/constants';

type DialogProps = React.PropsWithChildren<{
  open: boolean;
  onClose?(): void;
  className?: string;
}>;

export function Dialog(props: DialogProps) {
  return (
    <HeadlessDialog open={props.open} onClose={props.onClose} className={clsx(props.className, FULLSTORY_UNMASK)}>
      {props.children}
    </HeadlessDialog>
  );
}

```

So now I am forced to keep mapping things into the next object rather than using ESM, so I can't do the following:

```tsx
export { Overlay, Title } from '@headlessui/react';
```

I can either do imperative things, hopefully getting it right by copying over the keys, or I can leverage ESM for it which already has built-in capabilities for re-exporting things.

----

Any thoughts?

Worth saying, I know how to fix the issue, but I would like to propose to back track out of such a decision made, from my experience, it adds a lot of complexity and burden on the consumers.

**Comment by PedroAugustoRamalhoDuarte** [comment]
Hello @yordis, I am having trouble to make my component library tree-shakeble using headlessui. I think is a similar issue, do you know how I can achieve tree-shakable using Dialog component for example?
